[linter / doc] Add example parameter linter rule.

adewaleo opened the pull request:
Adds an example parameter linter rule to prevent issues like #8054, #8150 

Todo:
- [x] linter should not be aggressive on alias examples. E.g:
```
        az alias create --name 'list-vm {{ resource_group }}' \
          --command 'vm list --resource-group {{ resource_group }}'
```
- [x] linter should handle situations where a an argument to a parameter contains a non cli command. Eg:
```
az vm run-command invoke -g MyResourceGroup -n MyVm --command-id RunShellScript --scripts "sudo apt-get update && sudo apt-get install -y nginx"
```

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

adewaleo commented:
https://travis-ci.org/Azure/azure-cli/jobs/469327570#L964

There are currently 110 violations detected violations. Even if you account for rare false positives like those mentioned in the first comment (`alias create` and `vm run-command invoke`), there would still be a lot of violations!

tjprescott commented:
Thanks @adewaleo from the list of failures it seems like it may not accurately capture the `--foo=bar` style of specifying arguments, so please check that as well.

adewaleo commented:
> Thanks @adewaleo Tosin Adewale FTE from the list of failures it seems like it may not accurately capture the `--foo=bar` style of specifying arguments, so please check that as well.

I will address that @tjprescott. Thanks for the feedback.

williexu commented on tools/automation/cli_linter/rules/help_rules.py:
```diff
@@ -48,31 +49,36 @@ def faulty_help_example_rule(linter, help_entry):
+                    command_args, nested_commands = process_command_args(command_args)
+                    parser.parse_args(command_args)
+                except SystemExit as e:
+                    violation = "\tThe following command:\n\t\t{}\n\tis not a valid command".format(command)
```
I don't think you need to print "The following command...." for each command found. Just the command should be fine.

williexu commented on tools/automation/cli_linter/rules/help_rules.py:
```diff
@@ -48,31 +49,36 @@ def faulty_help_example_rule(linter, help_entry):
     if violations:
         num_err = len(violations)
-        violation_str = "\n".join(violations[:25])
+        violation_str = "\n".join(violations[:10])
```
Do we need to restrict the number of violations shown?

adewaleo replied:
I guess it is unnecessary. I will remove this

tjprescott replied:
I'm a little curious why the existing code limits the violations to 25?  I would find it annoying if it didn't display all of the problems in one go.

williexu commented on tools/automation/cli_linter/rules/help_rules.py:
```diff
@@ -48,31 +49,36 @@ def faulty_help_example_rule(linter, help_entry):
+        example_text = example.get('text','')
+        commands = extract_commands_from_example(example_text)
+        i = 0
+        while i < len(commands):
```
Why not just pop from commands and looping as long as it's not empty? 
No need to keep commands once theyve been parsed.

williexu commented on tools/automation/cli_linter/util.py:
```diff
@@ -89,39 +87,34 @@ class LinterError(Exception):
+def extract_commands_from_example(example_text):
+
+    # fold commands spanning multiple lines into one line. Split commands that use pipes
+    example_text = example_text.replace("\\\n", " ")
```
Will this not also combine separate commands into one?

adewaleo replied:
This Will not, it should only combine lines where the preceding line ends with a backslash and newline. To handle situations like [this](https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_help.py#L45-L56):

***
            az keyvault certificate create --vault-name vaultname -n cert1 \\
              -p "$(az keyvault certificate get-default-policy)"
            secrets=$(az keyvault secret list-versions --vault-name vaultname \\
              -n cert1 --query "[?attributes.enabled].id" -o tsv)
            vm_secrets=$(az vm secret format -s "$secrets")
            az vm create -g group-name -n vm-name --admin-username deploy  \\
              --image debian --secrets "$vm_secrets"
***

adewaleo commented:
@tjprescott @williexu, does everything look good to you? Please feel free to take a look at the CI and review the failing commands. We can discuss what to do with the incorrect examples in the new year.

tjprescott commented:
It looks like the output could be cleaned up, but otherwise it looks robust and very useful to me! Thanks @adewaleo!

adewaleo commented on src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_help.py:
```diff
@@ -526,16 +526,16 @@
-            az iot hub routing-endpoint create --resource-group MyResourceGroup --hub-name MyIotHub
-            --endpoint-name E2 --endpoint-type eventhub --endpoint-resource-group [Resource Group]
-            --endpoint-subscription-id [SubscriptionId] --connection-string [Connection String]
+          text: |
```
Changing from folded style to literal style was unnecessary, I can change it back. Although, if one wants to include backslashes properly I believe one would need to use block style. 

The only thing that was neccessary was to quote the arguments that had square brackets so they could be parsed properly by the CLIparser, if the value has spaces within it.

The parser also does not properly recognize braces, that have spaces in them. The solution is to either have no spaces in the option arguments / values or to quote such options.

We should also document somewhere that the syntax for specifying optionvalues in examples is indeed curly braces. Cause I don't believe we explicily mention this in our help authoring docs.

adewaleo replied:
Lol, fair enough. Hopefully #8231 would address this.

adewaleo commented on src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_help.py:
```diff
@@ -526,16 +526,16 @@
-            --endpoint-subscription-id [SubscriptionId] --connection-string [Connection String]
+          text: |
+            az iot hub routing-endpoint create --resource-group MyResourceGroup --hub-name MyIotHub \\
+            --endpoint-name E2 --endpoint-type eventhub --endpoint-resource-group "[Resource Group]" \\
```
Will make a pass to update any placeholders I changed to use curly braces

adewaleo replied:
I will also update our help authoring with example authoring guidelines and tips.

tjprescott commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -412,8 +412,8 @@
-                 -g <myRG> -n <myAppName> \\
-                 --src <zip file path location>
+                 -g {myRG} -n {myAppName} \\
+                 --src {zip_file_path_location}
```
This should use the same camelCase notation as the previous line.

tjprescott commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py:
```diff
@@ -993,7 +993,7 @@
              az functionapp deployment source config-zip \\
                  -g <myRG> -n <myAppName> \\
-                 --src <zip file path location>
+                 --src {zip_file_path_location}
```
Same here, both of these lines should use curly braces and camelCase notation.

tjprescott commented on src/command_modules/azure-cli-dms/azure/cli/command_modules/dms/_help.py:
```diff
@@ -44,7 +44,7 @@
         - name: Create an instance of DMS.
           text: >
-            az dms create -l westus -n mydms -g myresourcegroup --sku-name Basic_2vCores --subnet /subscriptions/{vnet subscription id}/resourceGroups/{vnet resource group}/providers/Microsoft.Network/virtualNetworks/{vnet name}/subnets/{subnet name} --tags tagName1=tagValue1 tagWithNoValue
+            az dms create -l westus -n mydms -g myresourcegroup --sku-name Basic_2vCores --subnet "/subscriptions/{vnet subscription id}/resourceGroups/{vnet resource group}/providers/Microsoft.Network/virtualNetworks/{vnet name}/subnets/{subnet name}" --tags tagName1=tagValue1 tagWithNoValue
```
Here I would replace the "things with spaces" with their "camelCaseEquivalent".  What error will users see if this is not quoted?

tjprescott commented on src/command_modules/azure-cli-hdinsight/azure/cli/command_modules/hdinsight/_help.py:
```diff
@@ -47,16 +47,16 @@
               az hdinsight application create -n MyCluster -g MyResourceGroup \\
               --application-name MyApplication \\
-              --script-uri https://path/to/install/script.sh
+              --script-uri "https://path/to/install/script.sh" \\
```
What error will the user see if this is not quoted?

tjprescott commented on src/command_modules/azure-cli-hdinsight/azure/cli/command_modules/hdinsight/_help.py:
```diff
@@ -47,16 +47,16 @@
+              --script-uri "https://path/to/install/script.sh" \\
               --script-action-name MyScriptAction \\
-              --script-parameters "-option value"
+              --script-parameters '"-option value"'
```
Why are these quotes necessary?

tjprescott replied:
The first instance looks correct to me, implying the second case of single quoting the double quotes is not necessary.  People should not have to add special formatting to the example to make the linter rule happy.

adewaleo replied:
The first instance fails with this error, because without quoting it, the cli itself treats it as an option, not a value....

```
(cli-venv) [05:59 PM] Work-Mac:azure-cli tosin$ az vm show -n "-option-value" -g ova-test
az vm show: error: argument --name/-n: expected one argument
usage: az vm show [-h] [--verbose] [--debug]
                  [--output {json,jsonc,table,tsv,yaml,none}]
                  [--query JMESPATH] [--resource-group RESOURCE_GROUP_NAME]
                  [--name NAME] [--show-details]
                  [--subscription _SUBSCRIPTION] [--ids ID [ID ...]]
(cli-venv) [05:59 PM] Work-Mac:azure-cli tosin$ 
```

adewaleo replied:
@tjprescott

tjprescott commented on src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_help.py:
```diff
@@ -624,7 +624,7 @@
         - name: Get an route information from "MyIotHub" IoT Hub.
           text: >
-            az iot hub route show -g MyResourceGroup --hub-name MyIotHub --route-name [Route Name]
+            az iot hub route show -g MyResourceGroup --hub-name MyIotHub --route-name "[Route Name]"
```
`{routeName}` without double quotes.

tjprescott commented on src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_help.py:
```diff
@@ -97,7 +97,7 @@
             -n rwallserviceaccess --validity-period P2D --sas-type account --template-uri $sastoken
         - name: Add a sas-definition for a blob sas-token
-          text: >
+          text: |
```
Why the switch?

tjprescott commented on src/command_modules/azure-cli-network/azure/cli/command_modules/network/_help.py:
```diff
@@ -90,11 +90,6 @@
 helps['network application-gateway update'] = """
     type: command
     short-summary: Update an application gateway.
-    examples:
```
Why was this simply removed?

adewaleo replied:
`--servers` is in the create command and not the update command and I did not want to create an app-gateway to confirm what a update --set command looks like. Maybe this could be a network-cli followup issue?

tjprescott commented on src/command_modules/azure-cli-network/azure/cli/command_modules/network/_help.py:
```diff
@@ -2583,7 +2577,7 @@
         - name: List all NICs by internal DNS suffix.
           text: >
-            az network nic list --query "[?dnsSettings.internalDomainNameSuffix=--query "[?dnsSettings.internalDomainNameSuffix==`<dns_suffix>`]"
+            az network nic list --query "[?dnsSettings.internalDomainNameSuffix=`<dns_suffix>`]"
```
`...internalDomanNameSuffix={dnsSuffix}`

tjprescott commented on src/command_modules/azure-cli-servicefabric/azure/cli/command_modules/servicefabric/_help.py:
```diff
@@ -167,7 +167,7 @@
         - name: Change the cluster durability level to 'Silver'.
           text: >
-            az sf cluster durability update -g group-name -n cluster1 --durability-level Silver
+            az sf cluster durability update -g group-name -n cluster1 --durability-level Silver --node-type 'ns1'
```
Why the single quotes on ns1? They aren't necessary.

adewaleo replied:
They are unnecessary, indeed. i used them, because I copied other examples.

tjprescott commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py:
```diff
@@ -86,7 +86,7 @@
           min_profile: latest
         - name: Create a storage account 'MyStorageAccount' in resource group 'MyResourceGroup' in the West US region with locally redundant storage.
-          text: az storage account create -n MyStorageAccount -g MyResourceGroup -l westus --account-type Standard_LRS
+          text: az storage account create -n MyStorageAccount -g MyResourceGroup -l westus --sku Standard_LRS
```
I believe this is actually incorrect. Note `max_profile: 2017-03-09-profile`. In the old storage create command, this was actually called `--account-type`. Does the rule check for profile dependency?

adewaleo commented:
@tjprescott @williexu, is this looking good?

adewaleo commented:
ping @tjprescott @williexu

williexu commented on tools/automation/cli_linter/rules/help_rules.py:
```diff
@@ -44,3 +47,109 @@ def faulty_help_example_rule(linter, help_entry):
+def _lint_example_command(command, parser, mocked_error_method, mocked_get_value, mocked_check_value):
+    def get_value_side_effect(action, arg_string):
+        return arg_string
+    mocked_error_method.side_effect = SystemExit  # mock call of parser.error so usage won't be printed.
```
Please do not use SystemExit as your side effect and create your own exception(or use https://github.com/Azure/azure-cli/blob/dev/tools/automation/cli_linter/util.py#L81)
This stops you from being able to differentiate between the side effect and other possible SystemExits.

adewaleo replied:
Makes sense, I will throw a LinterError

adewaleo commented on tools/automation/cli_linter/rules/help_rules.py:
```diff
@@ -44,3 +47,109 @@ def faulty_help_example_rule(linter, help_entry):
+
+@mock.patch("azure.cli.core.parser.AzCliCommandParser._check_value")
+@mock.patch("argparse.ArgumentParser._get_value")
+@mock.patch("azure.cli.core.parser.AzCliCommandParser.error")
```
The mocks mock out different parser behaviors. I mocked out check_value and get_value so that value validation does not fail. We don't want the parser to validate option values as they could be placeholders or incorrect or unassigned shell variables.

I mocked out AzCliCommandParser error to remove the side effects of the method when parsing fails. Side effect such as logging an error and printing the command's usage to stderr

adewaleo commented on tools/automation/cli_linter/rules/help_rules.py:
```diff
@@ -44,3 +47,109 @@ def faulty_help_example_rule(linter, help_entry):
+        violation_str = "\n\n".join(violations)
+        violation_msg = "\n\tThere is a violation:\n{}.".format(violation_str) if num_err == 1 else \
+            "\n\tThere are {} violations:\n{}".format(num_err, violation_str)
+        raise RuleError(violation_msg + "\n\n")
```
It looks like this:
```
Initializing linter with command table and help files...

	SKIPPING example: az storage account create -n MyStorageAccount -g MyResourceGroup -l westus --account-type Standard_LRS
	as its max profile is 2017-03-09-profile, instead of latest.
-  FAIL: faulty_help_example_parameters_rule
    Help-Entry: `acr build-task create` - 
	There is a violation:
	"az acr build-task create -t hello-world:{{.Build.ID}} -n hello-world -r MyRegistry -c https://github.com/Azure-Samples/acr-build-helloworld-node.git --commit-trigger-enabled false" is not a valid command.
	the following arguments are required: --git-access-token.


    Help-Entry: `acr build-task show-build` - 
	There is a violation:
	"az acr build-task show-build -n MyBuildTask -r MyRegistry --build-id aab1 -o table" is not a valid command.
	unrecognized arguments: -n MyBuildTask.


    Help-Entry: `acr task update` - 
	There is a violation:
	"az acr task update -n MyTask -r MyRegistry --base-image-trigger All --status Disabled" is not a valid command.
	ambiguous option: --base-image-trigger could match --base-image-trigger-enabled, --base-image-trigger-type.


    Help-Entry: `acs create` - 
	There are 2 violations:
	"az acs create -g MyResourceGroup -n MyContainerService --agent-profiles '["
	has no closing quotation. Tip: to continue an example command on the next line, use a "\" followed by a newline.
	If needed, you can escape the "\", like so "\\"

	"az acs create -g MyResourceGroup -n MyContainerService --agent-profiles '["
	has no closing quotation. Tip: to continue an example command on the next line, use a "\" followed by a newline.
	If needed, you can escape the "\", like so "\\"


    Help-Entry: `aks install-connector` - 
	There is a violation:
	"az aks install-connector --name MyManagedCluster --resource-group MyResourceGroup    --connector-name aci-connector --service-principal <SPN_ID> --client-secret <SPN_SECRET>    --aci-resource-group <ACI resource group>" is not a valid command.
	unrecognized arguments: resource group>.
```

adewaleo replied:
Success looks like this:
```
$ azdev cli-lint --rules faulty_help_example_parameters_rule
Initializing linter with command table and help files...

	SKIPPING example: az storage account create -n MyStorageAccount -g MyResourceGroup -l westus --account-type Standard_LRS
	as its max profile is 2017-03-09-profile, instead of latest.
-  pass: faulty_help_example_parameters_rule 
```

adewaleo commented:
Sorry, I assumed it was okay to dismiss the review, seeing I addressed your feedback @williexu,

I will wait till you review again / approve before merging.

adewaleo commented:
Work would be continued to lint help objects and not help entries, in preparation of the upcoming help overhaul. See #8266
